deflate.c

```diff
@@ -143,27 +143,17 @@ static const config configuration_table[10] = {
 
 // UPDATE_HASH(s,h,c) (h = (((h)<<s->hash_shift) ^ (c)) & s->hash_mask)
 
-static uint32_t hash_func_sse42(deflate_state *s, void* str) __attribute__ ((__target__ ("sse4.2")));
+static uint32_t hash_func_sse42(deflate_state *s, uint32_t h, void* str) __attribute__ ((__target__ ("sse4.2")));
 
-static uint32_t hash_func_sse42(deflate_state *s, void* str) {
+static uint32_t hash_func_sse42(deflate_state *s, uint32_t h, void* str) {
     return _mm_crc32_u32(0, *(uint32_t*)str) & s->hash_mask;
 }
 
-static uint32_t hash_func_default(deflate_state *s, void* str) {
-	uint32_t a = *(uint32_t*)str;
-// Thomas Wang
-	a += ~(a<<15);
-	a ^=  (a>>10);
-	a +=  (a<<3);
-	a ^=  (a>>6);
-	a += ~(a<<11);
-	a ^=  (a>>16);
-// -
-  a &= s->hash_mask;
-	return a;
+static uint32_t hash_func_default(deflate_state *s, uint32_t h, void* str) {
+	return ((h << s->hash_shift) ^ (*(uint32_t*)str)) & s->hash_mask;
 }
 
-static uint32_t hash_func(deflate_state *s, void* str) __attribute__ ((ifunc ("resolve_hash_func")));
```

Comment:
The function can't be inlined with ifunc. Did you measure the performance impact of that?